phat/crates/rollup/src/clients/ink.rs

```diff
@@ -0,0 +1,424 @@
+        })
+    }
+}
+
+impl<'a> KvSnapshot for InkSnapshot<'a> {
+    fn get(&self, key: &[u8]) -> kv_session::Result<Option<Value>> {
+        let contract = InkContract::new(self.rpc, self.pallet_id, self.call_id, self.contract_id);
+
+        // result of the query
+        type QueryResult = Option<Vec<u8>>;
+        // call the method
+        let value: QueryResult = contract
+            .query_at(
+                *self.contract_id,
+                ink::selector_bytes!("KvStore::get_value"),
+                Some(&key),
+                0,
+                Some(self.at),
+            )
+            .log_err("Rollup snapshot: failed to get storage")
+            .map_err(|_| kv_session::Error::FailedToGetStorage)?;
+
+        #[cfg(feature = "logging")]
+        debug!("Snapshot - key: {:02x?} - value: {:02x?}", &key, &value);
+
+        Ok(value)
+    }
+
+    fn snapshot_id(&self) -> kv_session::Result<Vec<u8>> {
+        Ok(self.at.encode())
+    }
+}
+
+impl<'a> BumpVersion for InkSnapshot<'a> {
+    fn bump_version(&self, version: Option<Vec<u8>>) -> kv_session::Result<Vec<u8>> {
+        match version {
+            Some(v) => {
+                let ver = u32::decode(&mut &v[..]).or(Err(kv_session::Error::FailedToDecode))?;
+                Ok((ver + 1).encode())
+            }
+            None => Ok(1u32.encode()),
+        }
+    }
+}
+
+pub struct ScaleCodec;
+impl QueueIndexCodec for ScaleCodec {
+    fn encode(number: QueueIndex) -> Vec<u8> {
+        number.encode()
+    }
+
+    fn decode(raw: impl AsRef<[u8]>) -> kv_session::Result<QueueIndex> {
+        // QueueIndex is stored as a value in the rollup kv store. When the value is empty, it's
+        // treated as the default value (0 for u32). However, this function only handles the
+        // non-empty case (empty value != zero length bytes). So here, `[]` is not considered.
+        QueueIndex::decode(&mut raw.as_ref()).or(Err(kv_session::Error::FailedToDecode))
+    }
+}
+
+pub struct InkRollupClient<'a> {
+    rpc: &'a str,
+    pallet_id: u8,
+    call_id: u8,
+    contract_id: &'a ContractId,
+    actions: Vec<Action>,
+    session: Session<InkSnapshot<'a>, RwTracker, ScaleCodec>,
+}
+
+pub struct SubmittableRollupTx<'a> {
+    rpc: &'a str,
+    pallet_id: u8,
+    call_id: u8,
+    contract_id: &'a ContractId,
+    tx: InkRollupTx,
+    _at: H256,
+}
+
+#[derive(Debug, Default, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct InkRollupTx {
+    conditions: Vec<(Key, Option<Value>)>,
+    updates: Vec<(Key, Option<Value>)>,
+    actions: Vec<Action>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode)]
+pub struct HandleActionInput {
+    pub action_type: u8,
+    pub action: Option<Vec<u8>>,
+    pub address: Option<ContractId>,
+    pub id: Option<QueueIndex>,
+}
+
+const ACTION_REPLY: u8 = 0;
+const ACTION_SET_QUEUE_HEAD: u8 = 1;
+
+impl Action {
+    fn encode_into_ink(self) -> HandleActionInput {
```

Comment:
As suggested, if you just use the original `enum Action` defined in the rollup crate, you don't need to convert the type manually

Reply:
I remember to think about it but never implemented it.
It is done now.